// @angular/compiler-cli/src/ngtsc/transform/src/declaration.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/transform/src/declaration", ["require", "exports", "tslib", "typescript", "@angular/compiler-cli/src/ngtsc/translator", "@angular/compiler-cli/src/ngtsc/transform/src/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var ts = require("typescript");
    var translator_1 = require("@angular/compiler-cli/src/ngtsc/translator");
    var utils_1 = require("@angular/compiler-cli/src/ngtsc/transform/src/utils");
    function declarationTransformFactory(compilation) {
        return function (context) {
            return function (fileOrBundle) {
                if (ts.isBundle(fileOrBundle)) {
                    // Only attempt to transform source files.
                    return fileOrBundle;
                }
                return compilation.transformedDtsFor(fileOrBundle, context);
            };
        };
    }
    exports.declarationTransformFactory = declarationTransformFactory;
    /**
     * Processes .d.ts file text and adds static field declarations, with types.
     */
    var DtsFileTransformer = /** @class */ (function () {
        function DtsFileTransformer(importRewriter, importPrefix) {
            this.importRewriter = importRewriter;
            this.ivyFields = new Map();
            this.imports = new translator_1.ImportManager(importRewriter, importPrefix);
        }
        /**
         * Track that a static field was added to the code for a class.
         */
        DtsFileTransformer.prototype.recordStaticField = function (name, decls) { this.ivyFields.set(name, decls); };
        /**
         * Transform the declaration file and add any declarations which were recorded.
         */
        DtsFileTransformer.prototype.transform = function (file, context) {
            var _this = this;
            var visitor = function (node) {
                // This class declaration needs to have fields added to it.
                if (ts.isClassDeclaration(node) && node.name !== undefined &&
                    _this.ivyFields.has(node.name.text)) {
                    var decls = _this.ivyFields.get(node.name.text);
                    var newMembers = decls.map(function (decl) {
                        var modifiers = [ts.createModifier(ts.SyntaxKind.StaticKeyword)];
                        var typeRef = translator_1.translateType(decl.type, _this.imports);
                        return ts.createProperty(undefined, modifiers, decl.name, undefined, typeRef, undefined);
                    });
                    return ts.updateClassDeclaration(node, node.decorators, node.modifiers, node.name, node.typeParameters, node.heritageClauses, tslib_1.__spread(node.members, newMembers));
                }
                // Otherwise return node as is.
                return ts.visitEachChild(node, visitor, context);
            };
            // Recursively scan through the AST and add all class members needed.
            var sf = ts.visitNode(file, visitor);
            // Add new imports for this file.
            return utils_1.addImports(this.imports, sf);
        };
        return DtsFileTransformer;
    }());
    exports.DtsFileTransformer = DtsFileTransformer;
});
