// @angular/compiler-cli/src/ngtsc/file_system/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system", ["require", "exports", "@angular/compiler-cli/src/ngtsc/file_system/src/cached_file_system", "@angular/compiler-cli/src/ngtsc/file_system/src/compiler_host", "@angular/compiler-cli/src/ngtsc/file_system/src/helpers", "@angular/compiler-cli/src/ngtsc/file_system/src/logical", "@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system", "@angular/compiler-cli/src/ngtsc/file_system/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var cached_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/cached_file_system");
    exports.CachedFileSystem = cached_file_system_1.CachedFileSystem;
    var compiler_host_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/compiler_host");
    exports.NgtscCompilerHost = compiler_host_1.NgtscCompilerHost;
    var helpers_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/helpers");
    exports.absoluteFrom = helpers_1.absoluteFrom;
    exports.absoluteFromSourceFile = helpers_1.absoluteFromSourceFile;
    exports.basename = helpers_1.basename;
    exports.dirname = helpers_1.dirname;
    exports.getFileSystem = helpers_1.getFileSystem;
    exports.isRoot = helpers_1.isRoot;
    exports.join = helpers_1.join;
    exports.relative = helpers_1.relative;
    exports.relativeFrom = helpers_1.relativeFrom;
    exports.resolve = helpers_1.resolve;
    exports.setFileSystem = helpers_1.setFileSystem;
    var logical_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/logical");
    exports.LogicalFileSystem = logical_1.LogicalFileSystem;
    exports.LogicalProjectPath = logical_1.LogicalProjectPath;
    var node_js_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/node_js_file_system");
    exports.NodeJSFileSystem = node_js_file_system_1.NodeJSFileSystem;
    var util_1 = require("@angular/compiler-cli/src/ngtsc/file_system/src/util");
    exports.getSourceFileOrError = util_1.getSourceFileOrError;
});
