// @angular/compiler-cli/ngcc/src/rendering/source_maps.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/rendering/source_maps", ["require", "exports", "convert-source-map", "source-map", "@angular/compiler-cli/src/ngtsc/file_system"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var convert_source_map_1 = require("convert-source-map");
    var source_map_1 = require("source-map");
    var file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system");
    /**
     * Get the map from the source (note whether it is inline or external)
     */
    function extractSourceMap(fs, logger, file) {
        var inline = convert_source_map_1.commentRegex.test(file.text);
        var external = convert_source_map_1.mapFileCommentRegex.exec(file.text);
        if (inline) {
            var inlineSourceMap = convert_source_map_1.fromSource(file.text);
            return {
                source: convert_source_map_1.removeComments(file.text).replace(/\n\n$/, '\n'),
                map: inlineSourceMap,
                isInline: true,
            };
        }
        else if (external) {
            var externalSourceMap = null;
            try {
                var fileName = external[1] || external[2];
                var filePath = file_system_1.resolve(file_system_1.dirname(file_system_1.absoluteFromSourceFile(file)), fileName);
                var mappingFile = fs.readFile(filePath);
                externalSourceMap = convert_source_map_1.fromJSON(mappingFile);
            }
            catch (e) {
                if (e.code === 'ENOENT') {
                    logger.warn("The external map file specified in the source code comment \"" + e.path + "\" was not found on the file system.");
                    var mapPath = file_system_1.absoluteFrom(file.fileName + '.map');
                    if (file_system_1.basename(e.path) !== file_system_1.basename(mapPath) && fs.exists(mapPath) &&
                        fs.stat(mapPath).isFile()) {
                        logger.warn("Guessing the map file name from the source file name: \"" + file_system_1.basename(mapPath) + "\"");
                        try {
                            externalSourceMap = convert_source_map_1.fromObject(JSON.parse(fs.readFile(mapPath)));
                        }
                        catch (e) {
                            logger.error(e);
                        }
                    }
                }
            }
            return {
                source: convert_source_map_1.removeMapFileComments(file.text).replace(/\n\n$/, '\n'),
                map: externalSourceMap,
                isInline: false,
            };
        }
        else {
            return { source: file.text, map: null, isInline: false };
        }
    }
    exports.extractSourceMap = extractSourceMap;
    /**
     * Merge the input and output source-maps, replacing the source-map comment in the output file
     * with an appropriate source-map comment pointing to the merged source-map.
     */
    function renderSourceAndMap(sourceFile, input, output) {
        var outputPath = file_system_1.absoluteFromSourceFile(sourceFile);
        var outputMapPath = file_system_1.absoluteFrom(outputPath + ".map");
        var relativeSourcePath = file_system_1.basename(outputPath);
        var relativeMapPath = relativeSourcePath + ".map";
        var outputMap = output.generateMap({
            source: outputPath,
            includeContent: true,
        });
        // we must set this after generation as magic string does "manipulation" on the path
        outputMap.file = relativeSourcePath;
        var mergedMap = mergeSourceMaps(input.map && input.map.toObject(), JSON.parse(outputMap.toString()));
        var result = [];
        if (input.isInline) {
            result.push({ path: outputPath, contents: output.toString() + "\n" + mergedMap.toComment() });
        }
        else {
            result.push({
                path: outputPath,
                contents: output.toString() + "\n" + convert_source_map_1.generateMapFileComment(relativeMapPath)
            });
            result.push({ path: outputMapPath, contents: mergedMap.toJSON() });
        }
        return result;
    }
    exports.renderSourceAndMap = renderSourceAndMap;
    /**
     * Merge the two specified source-maps into a single source-map that hides the intermediate
     * source-map.
     * E.g. Consider these mappings:
     *
     * ```
     * OLD_SRC -> OLD_MAP -> INTERMEDIATE_SRC -> NEW_MAP -> NEW_SRC
     * ```
     *
     * this will be replaced with:
     *
     * ```
     * OLD_SRC -> MERGED_MAP -> NEW_SRC
     * ```
     */
    function mergeSourceMaps(oldMap, newMap) {
        if (!oldMap) {
            return convert_source_map_1.fromObject(newMap);
        }
        var oldMapConsumer = new source_map_1.SourceMapConsumer(oldMap);
        var newMapConsumer = new source_map_1.SourceMapConsumer(newMap);
        var mergedMapGenerator = source_map_1.SourceMapGenerator.fromSourceMap(newMapConsumer);
        mergedMapGenerator.applySourceMap(oldMapConsumer);
        var merged = convert_source_map_1.fromJSON(mergedMapGenerator.toString());
        return merged;
    }
    exports.mergeSourceMaps = mergeSourceMaps;
});
