// @angular/compiler-cli/ngcc/src/migrations/undecorated_parent_migration.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/migrations/undecorated_parent_migration", ["require", "exports", "typescript", "@angular/compiler-cli/src/ngtsc/diagnostics", "@angular/compiler-cli/ngcc/src/utils", "@angular/compiler-cli/ngcc/src/migrations/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var ts = require("typescript");
    var diagnostics_1 = require("@angular/compiler-cli/src/ngtsc/diagnostics");
    var utils_1 = require("@angular/compiler-cli/ngcc/src/utils");
    var utils_2 = require("@angular/compiler-cli/ngcc/src/migrations/utils");
    /**
     * Ensure that the parents of directives and components that have no constructor are also decorated
     * as a `Directive`.
     *
     * Example:
     *
     * ```
     * export class BasePlain {
     *   constructor(private vcr: ViewContainerRef) {}
     * }
     *
     * @Directive({selector: '[blah]'})
     * export class DerivedDir extends BasePlain {}
     * ```
     *
     * When compiling `DerivedDir` which extends the undecorated `BasePlain` class, the compiler needs
     * to generate an `ngDirectiveDef` for `DerivedDir`. In particular, it needs to generate a factory
     * function that creates instances of `DerivedDir`.
     *
     * As `DerivedDir` has no constructor, the factory function for `DerivedDir` must delegate to the
     * factory function for `BasePlain`. But for this to work, `BasePlain` must have a factory function,
     * itself.
     *
     * This migration adds a `Directive` decorator to such undecorated parent classes, to ensure that
     * the compiler will create the necessary factory function.
     *
     * The resulting code looks like:
     *
     * ```
     * @Directive()
     * export class BasePlain {
     *   constructor(private vcr: ViewContainerRef) {}
     * }
     *
     * @Directive({selector: '[blah]'})
     * export class DerivedDir extends BasePlain {}
     * ```
     */
    var UndecoratedParentMigration = /** @class */ (function () {
        function UndecoratedParentMigration() {
        }
        UndecoratedParentMigration.prototype.apply = function (clazz, host) {
            // Only interested in `clazz` if it is a `Component` or a `Directive`,
            // and it has no constructor of its own.
            if (!utils_2.hasDirectiveDecorator(host, clazz) || utils_2.hasConstructor(host, clazz)) {
                return null;
            }
            // Only interested in `clazz` if it inherits from a base class.
            var baseClassExpr = host.reflectionHost.getBaseClassExpression(clazz);
            if (baseClassExpr === null) {
                return null;
            }
            if (!ts.isIdentifier(baseClassExpr)) {
                return diagnostics_1.makeDiagnostic(diagnostics_1.ErrorCode.NGCC_MIGRATION_EXTERNAL_BASE_CLASS, baseClassExpr, clazz.name.text + " class has a dynamic base class " + baseClassExpr.getText() + ", so it is not possible to migrate.");
                return null;
            }
            var baseClazz = host.reflectionHost.getDeclarationOfIdentifier(baseClassExpr).node;
            if (!utils_2.isClassDeclaration(baseClazz)) {
                return null;
            }
            // Only interested in this base class if it doesn't have a `Directive` or `Component` decorator.
            if (utils_2.hasDirectiveDecorator(host, baseClazz)) {
                return null;
            }
            var importInfo = host.reflectionHost.getImportOfIdentifier(baseClassExpr);
            if (importInfo !== null && !utils_1.isRelativePath(importInfo.from)) {
                return diagnostics_1.makeDiagnostic(diagnostics_1.ErrorCode.NGCC_MIGRATION_EXTERNAL_BASE_CLASS, baseClassExpr, 'The base class was imported from an external entry-point so we cannot add a directive to it.');
            }
            host.injectSyntheticDecorator(baseClazz, utils_2.createDirectiveDecorator(baseClazz));
            return null;
        };
        return UndecoratedParentMigration;
    }());
    exports.UndecoratedParentMigration = UndecoratedParentMigration;
});
