// @angular/compiler-cli/ngcc/src/migrations/utils.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/migrations/utils", ["require", "exports", "typescript", "@angular/compiler-cli/src/ngtsc/imports", "@angular/compiler-cli/src/ngtsc/reflection"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var ts = require("typescript");
    var imports_1 = require("@angular/compiler-cli/src/ngtsc/imports");
    var reflection_1 = require("@angular/compiler-cli/src/ngtsc/reflection");
    function isClassDeclaration(clazz) {
        return reflection_1.isNamedClassDeclaration(clazz) || reflection_1.isNamedFunctionDeclaration(clazz) ||
            reflection_1.isNamedVariableDeclaration(clazz);
    }
    exports.isClassDeclaration = isClassDeclaration;
    /**
     * Returns true if the `clazz` is decorated as a `Directive` or `Component`.
     */
    function hasDirectiveDecorator(host, clazz) {
        return host.metadata.getDirectiveMetadata(new imports_1.Reference(clazz)) !== null;
    }
    exports.hasDirectiveDecorator = hasDirectiveDecorator;
    /**
     * Returns true if the `clazz` has its own constructor function.
     */
    function hasConstructor(host, clazz) {
        return host.reflectionHost.getConstructorParameters(clazz) !== null;
    }
    exports.hasConstructor = hasConstructor;
    /**
     * Create an empty `Directive` decorator that will be associated with the `clazz`.
     */
    function createDirectiveDecorator(clazz) {
        var selectorArg = ts.createObjectLiteral([
            // TODO: At the moment ngtsc does not accept a directive with no selector
            ts.createPropertyAssignment('selector', ts.createStringLiteral('NGCC_DUMMY')),
        ]);
        var decoratorType = ts.createIdentifier('Directive');
        var decoratorNode = ts.createObjectLiteral([
            ts.createPropertyAssignment('type', decoratorType),
            ts.createPropertyAssignment('args', ts.createArrayLiteral([selectorArg])),
        ]);
        setParentPointers(clazz.getSourceFile(), decoratorNode);
        return {
            name: 'Directive',
            identifier: decoratorType,
            import: { name: 'Directive', from: '@angular/core' },
            node: decoratorNode,
            args: [selectorArg],
        };
    }
    exports.createDirectiveDecorator = createDirectiveDecorator;
    /**
     * Ensure that a tree of AST nodes have their parents wired up.
     */
    function setParentPointers(parent, child) {
        child.parent = parent;
        ts.forEachChild(child, function (grandchild) { return setParentPointers(child, grandchild); });
    }
    exports.setParentPointers = setParentPointers;
});
