// @angular/compiler-cli/src/transformers/tsc_pass_through.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/transformers/tsc_pass_through", ["require", "exports", "typescript", "@angular/compiler-cli/src/ngtsc/switch"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ts = require("typescript");
    var switch_1 = require("@angular/compiler-cli/src/ngtsc/switch");
    /**
     * An implementation of the `Program` API which behaves similarly to plain `tsc`.
     *
     * The only Angular specific behavior included in this `Program` is the operation of the Ivy
     * switch to turn on render3 behavior.
     *
     * This allows `ngc` to behave like `tsc` in cases where JIT code needs to be tested.
     */
    var TscPassThroughProgram = /** @class */ (function () {
        function TscPassThroughProgram(rootNames, options, host, oldProgram) {
            this.options = options;
            this.host = host;
            this.tsProgram =
                ts.createProgram(rootNames, options, host, oldProgram && oldProgram.getTsProgram());
        }
        TscPassThroughProgram.prototype.getTsProgram = function () { return this.tsProgram; };
        TscPassThroughProgram.prototype.getTsOptionDiagnostics = function (cancellationToken) {
            return this.tsProgram.getOptionsDiagnostics(cancellationToken);
        };
        TscPassThroughProgram.prototype.getNgOptionDiagnostics = function (cancellationToken) {
            return [];
        };
        TscPassThroughProgram.prototype.getTsSyntacticDiagnostics = function (sourceFile, cancellationToken) {
            return this.tsProgram.getSyntacticDiagnostics(sourceFile, cancellationToken);
        };
        TscPassThroughProgram.prototype.getNgStructuralDiagnostics = function (cancellationToken) {
            return [];
        };
        TscPassThroughProgram.prototype.getTsSemanticDiagnostics = function (sourceFile, cancellationToken) {
            return this.tsProgram.getSemanticDiagnostics(sourceFile, cancellationToken);
        };
        TscPassThroughProgram.prototype.getNgSemanticDiagnostics = function (fileName, cancellationToken) {
            return [];
        };
        TscPassThroughProgram.prototype.loadNgStructureAsync = function () { return Promise.resolve(); };
        TscPassThroughProgram.prototype.listLazyRoutes = function (entryRoute) {
            throw new Error('Method not implemented.');
        };
        TscPassThroughProgram.prototype.getLibrarySummaries = function () {
            throw new Error('Method not implemented.');
        };
        TscPassThroughProgram.prototype.getEmittedGeneratedFiles = function () {
            throw new Error('Method not implemented.');
        };
        TscPassThroughProgram.prototype.getEmittedSourceFiles = function () {
            throw new Error('Method not implemented.');
        };
        TscPassThroughProgram.prototype.emit = function (opts) {
            var emitCallback = opts && opts.emitCallback || defaultEmitCallback;
            var emitResult = emitCallback({
                program: this.tsProgram,
                host: this.host,
                options: this.options,
                emitOnlyDtsFiles: false,
                customTransformers: { before: [switch_1.ivySwitchTransform] },
            });
            return emitResult;
        };
        return TscPassThroughProgram;
    }());
    exports.TscPassThroughProgram = TscPassThroughProgram;
    var defaultEmitCallback = function (_a) {
        var program = _a.program, targetSourceFile = _a.targetSourceFile, writeFile = _a.writeFile, cancellationToken = _a.cancellationToken, emitOnlyDtsFiles = _a.emitOnlyDtsFiles, customTransformers = _a.customTransformers;
        return program.emit(targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers);
    };
});
