// @angular/compiler-cli/src/ngtsc/annotations/src/base_def.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/annotations/src/base_def", ["require", "exports", "tslib", "@angular/compiler", "@angular/compiler-cli/src/ngtsc/transform", "@angular/compiler-cli/src/ngtsc/annotations/src/directive", "@angular/compiler-cli/src/ngtsc/annotations/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var compiler_1 = require("@angular/compiler");
    var transform_1 = require("@angular/compiler-cli/src/ngtsc/transform");
    var directive_1 = require("@angular/compiler-cli/src/ngtsc/annotations/src/directive");
    var util_1 = require("@angular/compiler-cli/src/ngtsc/annotations/src/util");
    function containsNgTopLevelDecorator(decorators, isCore) {
        if (!decorators) {
            return false;
        }
        return decorators.some(function (decorator) { return util_1.isAngularDecorator(decorator, 'Component', isCore) ||
            util_1.isAngularDecorator(decorator, 'Directive', isCore) ||
            util_1.isAngularDecorator(decorator, 'NgModule', isCore); });
    }
    var BaseDefDecoratorHandler = /** @class */ (function () {
        function BaseDefDecoratorHandler(reflector, evaluator, isCore) {
            this.reflector = reflector;
            this.evaluator = evaluator;
            this.isCore = isCore;
            this.precedence = transform_1.HandlerPrecedence.WEAK;
        }
        BaseDefDecoratorHandler.prototype.detect = function (node, decorators) {
            var _this = this;
            if (containsNgTopLevelDecorator(decorators, this.isCore)) {
                // If the class is already decorated by @Component or @Directive let that
                // DecoratorHandler handle this. BaseDef is unnecessary.
                return undefined;
            }
            var result = undefined;
            this.reflector.getMembersOfClass(node).forEach(function (property) {
                var e_1, _a;
                var decorators = property.decorators;
                if (!decorators) {
                    return;
                }
                try {
                    for (var decorators_1 = tslib_1.__values(decorators), decorators_1_1 = decorators_1.next(); !decorators_1_1.done; decorators_1_1 = decorators_1.next()) {
                        var decorator = decorators_1_1.value;
                        if (util_1.isAngularDecorator(decorator, 'Input', _this.isCore)) {
                            result = result || {};
                            var inputs = result.inputs = result.inputs || [];
                            inputs.push({ decorator: decorator, property: property });
                        }
                        else if (util_1.isAngularDecorator(decorator, 'Output', _this.isCore)) {
                            result = result || {};
                            var outputs = result.outputs = result.outputs || [];
                            outputs.push({ decorator: decorator, property: property });
                        }
                        else if (util_1.isAngularDecorator(decorator, 'ViewChild', _this.isCore) ||
                            util_1.isAngularDecorator(decorator, 'ViewChildren', _this.isCore)) {
                            result = result || {};
                            var viewQueries = result.viewQueries = result.viewQueries || [];
                            viewQueries.push({ member: property, decorators: decorators });
                        }
                        else if (util_1.isAngularDecorator(decorator, 'ContentChild', _this.isCore) ||
                            util_1.isAngularDecorator(decorator, 'ContentChildren', _this.isCore)) {
                            result = result || {};
                            var queries = result.queries = result.queries || [];
                            queries.push({ member: property, decorators: decorators });
                        }
                        else if (util_1.isAngularDecorator(decorator, 'HostBinding', _this.isCore) ||
                            util_1.isAngularDecorator(decorator, 'HostListener', _this.isCore)) {
                            result = result || {};
                            var host = result.host = result.host || [];
                            host.push(property);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (decorators_1_1 && !decorators_1_1.done && (_a = decorators_1.return)) _a.call(decorators_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            });
            if (result !== undefined) {
                return {
                    metadata: result,
                    trigger: null,
                };
            }
            else {
                return undefined;
            }
        };
        BaseDefDecoratorHandler.prototype.analyze = function (node, metadata) {
            var _this = this;
            var analysis = {
                name: node.name.text,
                type: new compiler_1.WrappedNodeExpr(node.name),
                typeSourceSpan: compiler_1.EMPTY_SOURCE_SPAN,
            };
            if (metadata.inputs) {
                var inputs_1 = analysis.inputs = {};
                metadata.inputs.forEach(function (_a) {
                    var decorator = _a.decorator, property = _a.property;
                    var propName = property.name;
                    var args = decorator.args;
                    var value;
                    if (args && args.length > 0) {
                        var resolvedValue = _this.evaluator.evaluate(args[0]);
                        if (typeof resolvedValue !== 'string') {
                            throw new TypeError('Input alias does not resolve to a string value');
                        }
                        value = [resolvedValue, propName];
                    }
                    else {
                        value = propName;
                    }
                    inputs_1[propName] = value;
                });
            }
            if (metadata.outputs) {
                var outputs_1 = analysis.outputs = {};
                metadata.outputs.forEach(function (_a) {
                    var decorator = _a.decorator, property = _a.property;
                    var propName = property.name;
                    var args = decorator.args;
                    var value;
                    if (args && args.length > 0) {
                        var resolvedValue = _this.evaluator.evaluate(args[0]);
                        if (typeof resolvedValue !== 'string') {
                            throw new TypeError('Output alias does not resolve to a string value');
                        }
                        value = resolvedValue;
                    }
                    else {
                        value = propName;
                    }
                    outputs_1[propName] = value;
                });
            }
            if (metadata.viewQueries) {
                analysis.viewQueries =
                    directive_1.queriesFromFields(metadata.viewQueries, this.reflector, this.evaluator);
            }
            if (metadata.queries) {
                analysis.queries = directive_1.queriesFromFields(metadata.queries, this.reflector, this.evaluator);
            }
            if (metadata.host) {
                analysis.host = directive_1.extractHostBindings(metadata.host, this.evaluator, this.isCore ? undefined : '@angular/core');
            }
            return { analysis: analysis };
        };
        BaseDefDecoratorHandler.prototype.compile = function (node, analysis, pool) {
            var _a = compiler_1.compileBaseDefFromMetadata(analysis, pool, compiler_1.makeBindingParser()), expression = _a.expression, type = _a.type;
            return {
                name: 'ngBaseDef',
                initializer: expression, type: type,
                statements: [],
            };
        };
        return BaseDefDecoratorHandler;
    }());
    exports.BaseDefDecoratorHandler = BaseDefDecoratorHandler;
});
