// @angular/compiler-cli/ngcc/src/analysis/migration_host.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/analysis/migration_host", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/diagnostics", "@angular/compiler-cli/ngcc/src/analysis/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var diagnostics_1 = require("@angular/compiler-cli/src/ngtsc/diagnostics");
    var util_1 = require("@angular/compiler-cli/ngcc/src/analysis/util");
    /**
     * The standard implementation of `MigrationHost`, which is created by the
     * `DecorationAnalyzer`.
     */
    var DefaultMigrationHost = /** @class */ (function () {
        function DefaultMigrationHost(reflectionHost, metadata, evaluator, handlers, analyzedFiles) {
            this.reflectionHost = reflectionHost;
            this.metadata = metadata;
            this.evaluator = evaluator;
            this.handlers = handlers;
            this.analyzedFiles = analyzedFiles;
        }
        DefaultMigrationHost.prototype.injectSyntheticDecorator = function (clazz, decorator) {
            var classSymbol = this.reflectionHost.getClassSymbol(clazz);
            var newAnalyzedClass = util_1.analyzeDecorators(classSymbol, [decorator], this.handlers);
            if (newAnalyzedClass === null) {
                return;
            }
            var analyzedFile = getOrCreateAnalyzedFile(this.analyzedFiles, clazz.getSourceFile());
            var oldAnalyzedClass = analyzedFile.analyzedClasses.find(function (c) { return c.declaration === clazz; });
            if (oldAnalyzedClass === undefined) {
                analyzedFile.analyzedClasses.push(newAnalyzedClass);
            }
            else {
                mergeAnalyzedClasses(oldAnalyzedClass, newAnalyzedClass);
            }
        };
        return DefaultMigrationHost;
    }());
    exports.DefaultMigrationHost = DefaultMigrationHost;
    function getOrCreateAnalyzedFile(analyzedFiles, sourceFile) {
        var analyzedFile = analyzedFiles.find(function (file) { return file.sourceFile === sourceFile; });
        if (analyzedFile !== undefined) {
            return analyzedFile;
        }
        else {
            var newAnalyzedFile = { sourceFile: sourceFile, analyzedClasses: [] };
            analyzedFiles.push(newAnalyzedFile);
            return newAnalyzedFile;
        }
    }
    function mergeAnalyzedClasses(oldClass, newClass) {
        var e_1, _a, _b, _c;
        if (newClass.decorators !== null) {
            if (oldClass.decorators === null) {
                oldClass.decorators = newClass.decorators;
            }
            else {
                var _loop_1 = function (newDecorator) {
                    if (oldClass.decorators.some(function (d) { return d.name === newDecorator.name; })) {
                        throw new diagnostics_1.FatalDiagnosticError(diagnostics_1.ErrorCode.NGCC_MIGRATION_DECORATOR_INJECTION_ERROR, newClass.declaration, "Attempted to inject \"" + newDecorator.name + "\" decorator over a pre-existing decorator with the same name on the \"" + newClass.name + "\" class.");
                    }
                };
                try {
                    for (var _d = tslib_1.__values(newClass.decorators), _e = _d.next(); !_e.done; _e = _d.next()) {
                        var newDecorator = _e.value;
                        _loop_1(newDecorator);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                (_b = oldClass.decorators).push.apply(_b, tslib_1.__spread(newClass.decorators));
            }
        }
        if (newClass.diagnostics !== undefined) {
            if (oldClass.diagnostics === undefined) {
                oldClass.diagnostics = newClass.diagnostics;
            }
            else {
                (_c = oldClass.diagnostics).push.apply(_c, tslib_1.__spread(newClass.diagnostics));
            }
        }
    }
});
